transport/http/transport.go

```diff
@@ -174,16 +217,27 @@ func (a *Transport) RetainPeer(pid peer.Identifier, sub peer.Subscriber) (peer.P
 	return p, nil
 }
 
+// for httpPeer to obtain a peer or nil
+func (a *Transport) getPeerWithLock(addr string) *httpPeer {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	if p, ok := a.peers[addr]; ok {
+		return p
+	}
+	return nil
+}
+
 // **NOTE** should only be called while the lock write mutex is acquired
-func (a *Transport) getOrCreatePeer(pid hostport.PeerIdentifier) *hostport.Peer {
+func (a *Transport) getOrCreatePeer(pid hostport.PeerIdentifier) *httpPeer {
 	if p, ok := a.peers[pid.Identifier()]; ok {
 		return p
 	}
 
-	p := hostport.NewPeer(pid, a)
-	p.SetStatus(peer.Available)
-
+	p := newPeer(pid, a)
 	a.peers[p.Identifier()] = p
+	a.connectorsGroup.Add(1)
+	go p.maintainConn()
 
 	return p
```

Comment:
Consumers receive a notification when the availability of the peer changes. They are not obliged to heed that information, and can attempt to send a request immediately. The round robin and least pending heap implementations both heed those notifications by adding and removing the peer from the list of eligible peers and blocking peer selection until a peer surfaces.

All transports previously notified all peer lists immediately that all peers were implicitly available and the transport would attempt to create connections on demand. These changes increase the probability that availability is accurate.